python/examples/manual_message_decode.py

```diff
@@ -0,0 +1,131 @@
+
+    return True
+
+
+decode_message.expected_sequence_number = 0
+
+
+if __name__ == "__main__":
+    parser = ArgumentParser(description="""\
+Manually decode and print the contents of messages contained in a *.p1log file.
+
+This example application is similar to message_decode.py, but here we
+demonstrate manually decoding message headers and payloads, similar to what you
+might do for an incoming FusionEngine binary data stream.
+""")
+    parser.add_argument('--log-base-dir', metavar='DIR', default=DEFAULT_LOG_BASE_DIR,
+                        help="The base directory containing FusionEngine logs to be searched if a log pattern is "
+                             "specified.")
+    parser.add_argument('log',
+                        help="The log to be read. May be one of:\n"
+                             "- The path to a .p1log file or a file containing FusionEngine messages and other "
+                             "content\n"
+                             "- The path to a FusionEngine log directory\n"
+                             "- A pattern matching a FusionEngine log directory under the specified base directory "
+                             "(see find_fusion_engine_log() and --log-base-dir)")
+    options = parser.parse_args()
+
+    # Configure logging.
+    logging.basicConfig(format='%(message)s', stream=sys.stdout)
+    logger.setLevel(logging.INFO)
+
+    # Locate the input file and set the output directory.
+    #
+    # Note that, unlike the message_decode.py example, here we _do_ ask locate_log() to create a *.p1log file for us if
+    # it finds an input file containing a mix of FusionEngine messages and other data. The loop below assumes that the
+    # file we are reading contains _only_ FusionEngine messages.
+    input_path, output_dir, log_id = locate_log(options.log, return_output_dir=True, return_log_id=True,
+                                                log_base_dir=options.log_base_dir)
+    if input_path is None:
+        sys.exit(1)
+
+    if log_id is None:
+        logger.info('Loading %s.' % os.path.basename(input_path))
+    else:
+        logger.info('Loading %s from log %s.' % (os.path.basename(input_path), log_id))
+
+    # Read one FusionEngine message at a time from the binary file. If the file contains any data that is not part of a
+    # valid FusionEngine message, we will exit immediately.
+    f = open(input_path, 'rb')
+
+    while True:
+        # Read the next message header.
+        data = f.read(MessageHeader.calcsize())
+        if len(data) == 0:
+            break
+
+        # Deserialize the header.
+        try:
+            header = MessageHeader()
+            offset = header.unpack(buffer=data)
+        except Exception as e:
+            logger.error('Decode error: %s' % str(e))
+            break
+
+        # Read the message payload and append it to the header.
+        data += f.read(header.payload_size_bytes)
+
+        if not decode_message(header, data, offset):
```

Comment:
Minor thing, but most errors in decode_message (a crc or unpack error) will trigger an exception not return False. It would probably be more illustrative to catch those and return false.

Reply:
Yeah, good point